### PR TITLE
Feat/1st stage training script

### PR DESCRIPTION
**Pull request by deep-diver**
base code of training script for 1st stage SFT model

### CODE REVIEW

File: sft_1st_stage.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+def get_dataset(args):
+    dataset = load_dataset(args.dataset_name, split="train")
+    return dataset
+
+def run_training(args, dataset):
+    use_wandb = len(args.wandb_project) > 0 or (
+        "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
+    )
+    # Only overwrite environ if wandb param passed
+    if len(args.wandb_project) > 0:
```
Comment (sayakpaul) [inline sft_1st_stage.py:55]: I think it's better to default this CLI argument to None and condition on that. When am argument is provided then we log everything relevant. WDYT?

File: sft_1st_stage.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    
+    parser.add_argument("--wandb_project", type=str, default="")
+    parser.add_argument("--wandb_run_name", type=str, default="")
+    parser.add_argument("--wandb_watch", type=str, default="false") # options: false | gradients | all
+    parser.add_argument("--wandb_log_model", action="store_false")
+    
+    parser.add_argument("--lora_r", type=int, default=6)
+    parser.add_argument("--lora_alpha", type=int, default=32)
+    parser.add_argument("--lora_dropout", type=float, default=0.05)
+    parser.add_argument("--lora_target_modules", nargs="+", default=["c_proj", "c_attn", "q_attn"])
```
Comment (sayakpaul) [inline sft_1st_stage.py:42]: Using the linear modules for LoRA might worth trying too.

File: sft_1st_stage.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        train_dataset=dataset,
+        dataset_text_field=args.target_text_field,
+        max_seq_length=args.max_seq_length,
+        packing=args.packing,
+        peft_config=lora_config,
+        args=training_args,
+    )
+    trainer.train()
+    
+    trainer.model.save_pretrained(os.path.join(args.output_dir, "final_checkpoint/"))
```
Comment (sayakpaul) [inline sft_1st_stage.py:110]: Let's also have a utility for pushing to the HF hub. WDYT?

**Review by sayakpaul — APPROVED** [review]
LGTM with a few nits.

File: sft_1st_stage.py (hunk 2 above)
Comment (deep-diver) [reply sft_1st_stage.py:42]: I see there are couple of linear layers. which one do you have in mind?

I also noticed `q_attn` does not exist. my bad

```
GPTBigCodeForCausalLM(
  (transformer): GPTBigCodeModel(
    (wte): Embedding(49152, 6144)
    (wpe): Embedding(8192, 6144)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-39): 40 x GPTBigCodeBlock(
        (ln_1): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (attn): GPTBigCodeAttention(
          (c_attn): Linear(in_features=6144, out_features=6400, bias=True)
          (c_proj): Linear(in_features=6144, out_features=6144, bias=True)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTBigCodeMLP(
          (c_fc): Linear(in_features=6144, out_features=24576, bias=True)
          (c_proj): Linear(in_features=24576, out_features=6144, bias=True)
          (act): GELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=6144, out_features=49152, bias=False)
)
```

**Comment by deep-diver** [comment]
addressed your comments @sayakpaul 

about pushing the model to hub, I added relevant arguments to `TrainingArgument`, `push_to_hub`, `hub_model_id`, and `hub_token`.

about the comment below,
> Using the linear modules for LoRA might worth trying too.

You mean you want to include them by default?

File: sft_1st_stage.py (hunk 2 above)
Comment (sayakpaul) [reply sft_1st_stage.py:42]: Pretty cool thread here: https://twitter.com/mrm8488/status/1672355317487640577?s=20

File: sft_1st_stage.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    use_wandb = args.wandb_project or (
+        "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
+    )
+    # Only overwrite environ if wandb param passed
+    if args.wandb_project:
+        os.environ["WANDB_PROJECT"] = args.wandb_project
+    if len(args.wandb_watch) > 0:
+        os.environ["WANDB_WATCH"] = args.wandb_watch
+    if args.wandb_log_model:
+        os.environ["WANDB_LOG_MODEL"] = str(args.wandb_log_model)
```
Comment (sayakpaul) [inline sft_1st_stage.py:64]: Shouldn't this be conditioned on `if args.wandb_project:`? Also, I would be more explicit - `if args.wandb_project is not None:` (nit).

File: sft_1st_stage.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    
+    if args.push_to_hub:
+        if not hub_model_id:
+            print("You set `push_to_hub=True`, but `hub_model_id` is not provided.")
```
Comment (sayakpaul) [inline sft_1st_stage.py:68]: For the `hub_model_id` we could do something like:

https://github.com/huggingface/diffusers/blob/61916fefc400d1b77171490b6ff7d53e509ed9de/examples/dreambooth/train_dreambooth_lora.py#L745

Then we could remove the `hub_token` argument easily and just instruct the user that they should run `huggingface-cli login` if `push_to_hub` is True.

WDYT?

**Review by sayakpaul — COMMENTED** [review]
Some more comments :)

I can add utilities for pushing the adapter weights to the Hub and for running some validation inference during training to track progress as we do here: https://github.com/huggingface/diffusers/blob/61916fefc400d1b77171490b6ff7d53e509ed9de/examples/dreambooth/train_dreambooth_lora.py#L1253.

File: sft_1st_stage.py (hunk 2 above)
Comment (deep-diver) [reply sft_1st_stage.py:42]: so, in this architecture, `c_attn, `c_proj`, `c_fc`, and `lm_head` could be applied

**Comment by deep-diver** [comment]
@sayakpaul 
> I can add utilities for pushing the adapter weights to the Hub and for running some validation inference during training to track progress as we do here

that would be great! let's do that after merging this PR!

File: sft_1st_stage.py (hunk 5 above)
Comment (deep-diver) [reply sft_1st_stage.py:69]: Could you elaborate? The reference you shared seems like using custom `create_repo` function, and it still requires `token` argument.

**Review by sayakpaul — APPROVED** [review]
Thanks a lot for iterating.

**Pull request merged**